pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java

```diff
@@ -257,47 +390,162 @@ private void setTargetState(IdealState idealState, String segmentId, Map<String,
+        prettyPrintMapDebug(mapEV);
+      }
+
       for (String server : mapIS.keySet()) {
-        String state = mapIS.get(server);
-        if (mapEV == null || mapEV.get(server) == null || !mapEV.get(server).equals(state)) {
-          LOGGER.debug("Mismatch: segment" + segment + " server:" + server + " state:" + state);
-          numDiff = numDiff + 1;
+        if (!mapEV.containsKey(server)) {
+          LOGGER.debug("Host-state mapping of segment {} doesn't yet have server {} in external view",
+              segment, server);
+          // external view not yet converged
+          stable = false;
+        } else if (mapEV.get(server).equalsIgnoreCase("error")) {
+          LOGGER.error("Detected error state for segment {} for server {}", segment, server);
+          prettyPrintMapError(mapIS);
+          prettyPrintMapError(mapEV);
+          throw new IllegalStateException("External view reports error state for segment " + segment + " for host " + server,
+              new ExternalViewErrored());
+        } else {
+          final String stateInIdealState = mapIS.get(server);
+          final String stateInExternalView = mapEV.get(server);
+          if (!stateInIdealState.equalsIgnoreCase(stateInExternalView)) {
+            LOGGER.debug("Host-state mapping of segment {} has state {} in external view and state {} in ideal state",
+                segment, stateInExternalView, stateInIdealState);
+            // external view not yet converged
+            stable = false;
+          }
         }
       }
     }
-    return numDiff;
+    return stable;
+  }
+
+  private static void prettyPrintMapDebug(final Map<String, String> map) {
+    final Joiner.MapJoiner mapJoiner = Joiner.on(",").withKeyValueSeparator(":");
+    LOGGER.debug(mapJoiner.join(map));
+  }
+
+  private static void prettyPrintMapError(final Map<String, String> map) {
+    final Joiner.MapJoiner mapJoiner = Joiner.on(",").withKeyValueSeparator(":");
+    LOGGER.error(mapJoiner.join(map));
   }
 
   /**
    * Wait till state has stabilized {@link #isStable(String)}
    */
-  private void waitForStable(String resourceName)
-      throws InterruptedException {
-    int diff;
+  private void waitForStable(final String resourceName) {
     int INITIAL_WAIT_MS = 3000;
-    Thread.sleep(INITIAL_WAIT_MS);
-    do {
-      diff = isStable(resourceName);
-      if (diff == 0) {
-        break;
-      } else {
-        LOGGER.info(
-            "Waiting for externalView to match idealstate for table:" + resourceName + " Num segments difference:"
-                + diff);
-        Thread.sleep(EXTERNAL_VIEW_CHECK_INTERVAL_MS);
-      }
-    } while (diff > 0);
+    try {
+      Thread.sleep(INITIAL_WAIT_MS);
+      while (true) {
+        if (isStable(resourceName)) {
+          break;
+        } else {
+          LOGGER.info("Waiting for externalView to match idealstate for table:" + resourceName);
+          Thread.sleep(EXTERNAL_VIEW_CHECK_INTERVAL_MS);
```

Comment:
Although we now bail out on detecting error in external view, there is still a possibility of rebalancer waiting indefinitely for external view to converge simply because the changes haven't yet reflected and there is no error. I feel after updating ideal state in ZK when we come here to wait for stabilization, we should spend a fix amount of time in this method if isStable() never returns true

Reply:
agreed. Fixed, max time and the bail out.

Reply:
Done. using a max wait time of 5mins.